django/contrib/admin/views/autocomplete.py

```diff
@@ -0,0 +1,167 @@
+        considering the performance trade offs.
+        """
+        self.object_list.query.clear_ordering(force_empty=True)
+
+    def _get_page(self, *args, **kwargs):
+        return AutocompletePage(*args, **kwargs)
+
+    def validate_number(self, number):
+        """Validates the given 1-based page number but doesn't care about empty pages."""
+        try:
+            number = int(number)
+        except (TypeError, ValueError):
+            raise PageNotAnInteger('That page number is not an integer')
+        if number < 1:
+            raise EmptyPage('That page number is less than 1')
+        return number
+
+    @cached_property
+    def count(self):
+        """
+        Do NOT count the result set.
+
+        A count() or exist() filtered query can be very slow on large tables.
+        This means Select2 will do one more request that it needs to, but
+        that is a small trade off.
+        """
+        return sys.maxsize
+
+
+class AutocompleteJsonView(BaseListView):
+    """
+    View that handles requests from `.AutocompleteWidget`.
+
+    The view only supports HTTP's GET method.
+    """
+    paginator_class = AutocompletePaginator
+    paginate_by = 20
+    admin_site = None
+
+    @property
+    def target_model(self):
+        return self.source_field.remote_field.target_model
+
+    @property
+    def source_field(self):
+        return self.source_model._meta.get_field(self.field_name)
+
+    @property
+    def target_model_admin(self):
+        try:
+            return self.admin_site._registry[self.target_model]
+        except KeyError:
+            raise Http404('%s is not registered in the admin.' % self.target_model)
+
+    @property
+    def source_model_admin(self):
+        return self.admin_site._registry[self.source_model]
+
+    def has_perm(self, request):
+        """Check if user has permission to access the related model."""
+        return self.source_model_admin.has_change_permission(request)
+
+    def get(self, request, *args, **kwargs):
+        """
+        Return JSON search result based on a field identifier and a search term.
+
+        Returns:
+             (django.http.JsonResponse)::
+                {
+                    results: [
+                        {
+                            id: "123"
+                            text: "foo",
+                        }
+                    ],
+                    pagination: {
+                        more: true
+                    }
+                }
+        """
+        from django.apps import apps
+
+        self.app_label = kwargs.get('app_label')
+        self.model_name = kwargs.get('model_name')
+        self.field_name = kwargs.get('field_name')
+        self.term = request.GET.get('term', '')
+        self.source_model = apps.get_model(self.app_label, self.model_name)
+
+        # Since the list of registered autocomplete fields is generated during the request
```

Comment:
@apollo13 had a similar concern. There is not fix for that unless is discard @timgraham request for the `get_autocomplete_fields`.
The url routes are build during startup, where the autocomplete fields are not defined until the very request.

I can exclude it from the url pattern and throw it into the GET request. It will be only a different solution to the same problem.

Whatever you think is best really.